proposals/0000-extract-runtime.md

```diff
@@ -0,0 +1,419 @@
+...
+```
+
+Integration with another languages can be achieved by leveraging C++ API or creating a C API compatible
+API subset. It is, however, out of scope for this RFC.
+
+### Runtime Composability
+
+The goal of runtime composability is to essentially provide the features you need without paying for
+ones you do not use.
+
+This can be achieved by updating the process of how TurboModules are discovered and handled at build time level.
+Considering these TurboModules aim to expose new APIs to the JS world, we could make them conditionally linked
+(either statically or dynamically).
+
+There are two approaches we can further discover:
+
+ - Leveraging dynamic library constructor and destructor functions (`attribute(constructor)`/`attribute(destructor)`)
+   
+   This can be hard to do right, but ideally would be build-time only without any changes required in code or at runtime.
+   Unfortunately this only work with dynamic libraries, so we would require all modules to be linked dynamically which is suboptimal.
+  
+ - Exposing a Well known function that registers itself in React Native runtime 
+   
+   Similar to what we are doing now with TurboModules, we expect the library to export a function that we call in order to 
+   augment to runtime. This is probably more compatible solution but also requires having a collection of expected 
+   modules in code in order to load them.
+
+Application develop should be able to decide which of the libraries should be linked. React Native would then
+depend on the built runtime library. Both static linking and dynamic linking should be supported.
+
+### Per-Platform implementation
+
+Suggested approach still allows developers to have runtime level TurboModules use platform specific code
+by linking different libraries depending on the target platform.
+
+Leveraging the "take what you need only" approach allows to tailor the final application bundle just as needed
+for each platform. This, however requires additional work from the build tool.
+
+An example of such library that requires per-platform implementation and also is UI independent is `react-native-mmkv`.
+It implements a generic, headless API that hooks up to correct storage depending on the platform.
+
+A potential approach for per-platform implementation is to split the libraries and user C++ interop:
+
+```
+├── lib-rnrt-mmkv-core
+├── lib-rnrt-mmkv-android (JNI)
+```
+
+This can result in (subjectively healthy) distinction of platform specific code that either depends on UI or not.
+Consider Android and iOS platforms being split into:
+
+ - Android
+   - `lib-android`
+   - `lib-android-ui`
+ - iOS
+    - `lib-ios` (or perhaps `lib-apple`, as this is UI independent)
+    - `lib-ios-ui`
+
+This would benefit OOT platforms where only UI needs to be changed. 
+
+### Faster iterations & versioning
+
+Having `react-native-runtime` and `react-native` being versioned and released independently allows for simpler updating in
```

Comment:
realistically the core runtime would be released in sync with react-native because it would be a part of the monorepo. Unless you're thinking about a user-owned runtime implementation that's independently published. In such a case I'd make it obvious where the `react-native-runtime` lives

Reply:
Elaborated on this point